First Pass at Fully Implementing the PropertyEstimator Calls

SimonBoothroyd opened the pull request:
## Description
This PR is the result of a first pass of integrating the propertyestimator into the template provided by Yudong.

## Status
 - [x] Ready to go

Commit:
First pass at fully implementing the PE calls.
Cleaned up the config dictionary into a full object.
Minor reformatting and code renaming to be more comprehensible.

Commit:
Finished adding support for calculating gradients not using the PE API.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+    import propertyestimator
 except ImportError:
-    warn_once("Failed to import property_estimator package.")
+    # TODO: Shouldn't this be allowed to raise an exception,
```
We want to be a bit careful that people may get ForceBalance crash even when they do not use the property estimator at all.
In long term we may have a list of optional dependencies that people can install when they need specific features like this.

yudongqiu replied:
Therefore it would be nice to move these imports from above into this try block:
```
from openforcefield.typing.engines import smirnoff
from propertyestimator.client import PropertyEstimatorClient, ConnectionOptions, PropertyEstimatorOptions
from propertyestimator.datasets import ThermoMLDataSet
from propertyestimator.utils.exceptions import PropertyEstimatorException
from simtk import unit
```

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
-        1. Read input file
-        2. create client
-        3. use client to get information about the experimental dataset
+    class OptionsFile:
```
I think it's better to put these structured classes in the property estimator client. This have multiple benefits:
1. Other people using the interface to propertyestimator will be able to reuse the code instead of implementing themselves.
2. The configuration classes are expected to be updated with the propertyestimator client, as part of the API. Leaving them here will require you to constantly create PRs to the ForceBalance.

SimonBoothroyd replied:
The property estimator does not (and there are currently no plans to) support using these JSON config files for launching estimates. 

My understanding is that you wanted to be able to have a user configurable way to choose the data sets and property weights for the property estimator target, which these classes facilitate in a stable way (as opposed to arbitrary options dicts). But these are very much FB specific and as such should live with FB.

yudongqiu replied:
Sorry I think there's a bit confusion here. My understanding is that we have two sets of options.
One set is specific to property estimator, for the `client_config` and provide the data source as a list of doi's.
Second set of options is specific to ForceBalance, that contains the weights and denominators for each property.
I guess `propertyestimator` should have a simply way to load the first set of options in, without asking the user to implement a nested classes structure.
For the second set of options I agree it's ForceBalance specific, so I intend to leave it in a separate field, or even in a separate file.

SimonBoothroyd replied:
I should clarify that the `propertyestimator` does not load in a set of nested options. The client object only has a simple [programatic API](https://property-estimator.readthedocs.io/en/latest/api/generated/propertyestimator.client.PropertyEstimatorClient.html#propertyestimator.client.PropertyEstimatorClient) which accepts an optional `ConnectionOptions` argument - there is no support for loading in options from JSON. Users of the `propertyestimator` are expected to use the API directly in a fashion similar to the linked to example.

The nested class structure here is solely to act as a programatic object model for the JSON file stored with FB to help with code linting (i.e. you could remove these classes entirely and just use the dict loaded from JSON, but I wouldn't recommend it for code maintainability) - a user will never have to use these. The JSON file itself is very much a FB construct and a way for FB users to specify which properties to use with the target, as well as optional options to specify which property estimator server to connect to. These options represent a specific use case which in general may not be applicable to the `propertyestimator` in general, and it's up to the FB target to turn these user specified options into property estimator API calls so I still think these options belong here.

yudongqiu replied:
Sure let's keep them here for now. I may be able to simplify that a bit in the future 👍

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+
+    # A mapping of property estimator property names to
+    # force balance names.
+    estimator_to_force_balance_map = {
```
This is not needed. ForceBalance do not need the specific names of the properties, but should get them directly from the property estimator API.

yudongqiu replied:
In other words, we can assume these property names like "Density" to be consistent in the input JSON file and in property estimator API.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+    }
+
+    # A dictionary of the units that force balance expects each property in.
+    default_units = {
```
The json input file should contain a denominator that removes the unit from the physical properties. ForceBalance can then work with the values of them. It could be nice to note the units of the denominators in the json file.

SimonBoothroyd replied:
Ah, I wondered what their purpose was. Aren't these less of a user choice and more of a 'FB expects these values in these units?' in which case it might be better for them to live in code (unless they should be user configurable?).

yudongqiu replied:
I agree. I think it's easier to require the values in a unit, instead of letting the user to specify his own unit. We should keep a note for the required units in the source code as well as in the input file.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+        into a force balance key-value pair.
+
+        TODO: Currently nothing actually stores the substance for
+              which a property was calculated (e.g. water, methanol)
```
I agree. The dictionary structure I implemented assumes the same format as current ForceBalance liquid target, that each molecule to be fitted is in a separate target. If it's more useful to group all molecules into one `PropertyEstimate_SMIRNOFF` target, we can add one more layer on top of for the data.

yudongqiu replied:
To be clear, the current layered data structure is: 
```property (e.g. density) -> phase points (298.15k, 1.0atm)```.
If we want to support multiple molecules, we can have 
```molecule (e.g. water) -> property (e.g. density) -> phase points (298.15k, 1.0atm)```

SimonBoothroyd replied:
That's fine if that's what FB expects, but then the target should be passed the molecule being fitted (either through the JSON config) or an attribute. This will also determine which property doi's should be chosen (we can just filter the full data set based on the molecule if needed).

SimonBoothroyd replied:
Happy to go either route depending on what is better for FB :)

yudongqiu replied:
Sure. If we start with a list of DOIs and they contain multiple molecules, we could provide an additional filter to `ThermoMLDataSet` about what molecule and what property is needed from these DOIs. By default I assume we would want all possible data.

yudongqiu replied:
Because ForceBalance does not know what molecules and properties are available, it would require the user to run it twice to specify this filter.
In the first run, the user can provide just the DOIs, and when this target is created, it will send the DOIs to `ThermoMLDataSet` and print the names of the molecules and properties. If the user does not interrupt, this run will continue with fitting all available molecules and properties supported by the forcefield.
If the user needs to specify a filter, then he can interrupt the first run, copy and paste the names of the molecules and the properties into the input file, and this time `propertyestimator` will only compute the molecules and properties passes the filter.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+
+        # Load in the options from a user provided JSON file.
+        with open(os.path.join(self.tgtdir, self.prop_est_input), 'r') as file:
+            self._options = self.OptionsFile.from_json(file.read())
```
When we have the `OptionsFile` included in the `propertyestimator` API, this could become
```
prop_est_input_dict = json.load(json_file)
property_estimator_options_dict = prop_est_input_dict['property_estimator_options']
self._options = PropertyEstimatorOptions.from_dict(property_estimator_options_dict)
property_estimator_client_config = prop_est_input_dict['client_config']
connection_options = ConnectionOptions.from_dict(property_estimator_client_config)
```

yudongqiu replied:
My understanding is that you can customize the JSON input file to fit all the needs of the property estimator. The only data ForceBalance need is the weights and denominators, which can be kept in the JSON file as an additional field, or even in a separate file.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+        self._data_set = ThermoMLDataSet.from_file(*[source.path for source in
+                                                     self._options.sources if source.path])
+
+        self._data_set.merge(ThermoMLDataSet.from_doi(*[source.doi for source in
```
This is similar to what I guessed. I'm curious that how does `ThermoMLDataSet` extract data from the doi. Does it connects to an external database?

SimonBoothroyd replied:
It uses the doi to pull from the ThermoML archive, e.g. https://trc.nist.gov/ThermoML/10.1021/acs.jced.6b00916.xml

yudongqiu replied:
Sure thanks. So I guess it downloads the corresponding xml file from the archive and parse the values. This introduce an extra requirement that the node running this should have stable internet connection.
I also saw that it can load the data from file directly, which is very nice for testing.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,401 @@
+
+            for physical_property in self._data_set.properties[substance_identifier]:
+
+                class_name = physical_property.__class__.__name__
```
Same as above, that we can skip the mapping and directly use the names defined in propertyestimator. It would be nice if we have `physical_property.property_name` instead of `physical_property.__class__.__name__`

SimonBoothroyd replied:
It may look a bit ugly but the property name _is_ the name of the class and unfortunately this is the standard way of accessing such.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -142,59 +416,169 @@ def submit_jobs(self, mvals, AGrad=True, AHess=True):
+                                   for result in results.exceptions)
+
+            # TODO: How should we handle this case.
+            raise ValueError(f'Some properties could not be estimated:\n\n{exceptions}.')
```
I think raising an error is good here to let the user know the property estimation failed.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -142,59 +416,169 @@ def submit_jobs(self, mvals, AGrad=True, AHess=True):
+            # In some cases, an exception will be raised when executing a property but
+            # it will not stop the property from being estimated (e.g an error occured
+            # while reweighting so a simulation was used to estimate the property instead).
+            logger.warning(f'A number of non-fatal exceptions occured:\n\n{exceptions}')
```
Looks great.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -142,59 +416,169 @@ def submit_jobs(self, mvals, AGrad=True, AHess=True):
-        job_id = '1293791'
-        return job_id
+
+        # TODO: A given property may have been estimated for a number of different substances.
```
As suggested above, the substances could be at the top layer, and the property can be at the second layer. The same property for difference substances then will be stored in separate places.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -206,74 +590,17 @@ def wq_complete(self):
+            for direction in self._pending_gradient_requests[parameter_index]:
 
+                request = self._pending_gradient_requests[parameter_index][direction]
+                all_finished = all_finished & self._is_request_finished(request)
```
We could break early here if `all_finished` become False.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -303,67 +629,93 @@ def get(self, mvals, AGrad=True, AHess=True):
+        assert len(self._pending_gradient_requests) == self.FF.np, 'number of submitted jobs not consistent'
+
+        zero_gradient = np.zeros(self.FF.np)
+        estimated_gradients = {}
```
The gradients calculation can be skipped when `AGrad != True`

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -303,67 +629,93 @@ def get(self, mvals, AGrad=True, AHess=True):
+
+        for property_name in self._reference_properties:
+
             self.last_obj_details[property_name] = []
```
`self.last_obj_details` has been renamed to `self._last_obj_details` above

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -303,67 +629,93 @@ def get(self, mvals, AGrad=True, AHess=True):
-                tar_value = prop_est_data['values'][property_name][(temperature, pressure)]
-                tar_error = prop_est_data['value_errors'][property_name][(temperature, pressure)]
+
+                # TODO: load computed value and standard error from real data - is this done now?
```
Yes I think it's good.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -303,67 +629,93 @@ def get(self, mvals, AGrad=True, AHess=True):
                 obj_value += obj_contrib
-                self.last_obj_details[property_name].append((temperature, pressure, ref_value, tar_value, tar_error, diff, weight, denom, obj_contrib))
+
+                self.last_obj_details[property_name].append(
```
self.last_obj_details has been renamed to self._last_obj_details above

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -303,67 +629,93 @@ def get(self, mvals, AGrad=True, AHess=True):
+                    grad_array = estimated_gradients[property_name][phase_point_key]
                     # compute objective gradient
-                    this_obj_grad = 2.0 * weight * diff * grad_array / denom**2
+                    # TODO: `this_obj_grad` is not used?
```
Sorry for the bug. `this_obj_grad` should be accumuated to `ojb_grad` like this:
```
ojb_grad += this_obj_grad
```

yudongqiu commented:
Thanks a lot for the PR. I think the changes overall looks great, and a few minor issues can be fixed quickly. In general, I have two suggestions:

- [x] Customize the input file and simplify the API calls to property estimator.

- [ ] Reproduce the ForceBalance run in `studies/023_property_estimator`. (Sorry that I didn't include the targets folder before, I just committed a compressed file `targets.tar.gz` there).

Commit:
Added support for substances.
Made gradient calculation optional.
Refactored code to reduce redundency.
Temporarily made the estimation code blocking.
Turned off uncertainty checks.
Added support for JSON property files.
Moved imports into try / except block.

Commit:
Temporarily hardcoded the density convergence criteria.

Commit:
Fixed density criteria syntax.

Commit:
Added separate simulation / reweighting uncertainties.

Commit:
Fixed missing index.

Commit:
Added debug logging code.

Commit:
Added extra debug logging code.

Commit:
Fixed bad index access.

Commit:
Added file output for debugging.

Commit:
Temporarily made gradient requests blocking.

Commit:
Fix indicate method.

Commit:
Added debug logging info for gradients.

Commit:
Added extra debug logging info for gradients.

Commit:
Added extra debug information.

Commit:
Added extra debug information.

Commit:
Enforce that AGrad and AHess are bool type.

Commit:
Minor code refactor / cleanup.

leeping committed:
Merge pull request #145 from yudongqiu/master

fix small bug in smirnoffio.py

Commit:
Merge pull request #1 from yudongqiu/property_estimator

Sync Fork with Upstream

yudongqiu committed:
major improvements for smirnoff and related components

Commit:
First pass at implementing the mapping between pvals and mvals.

Commit:
Removed debug code.

Commit:
Fixed typo.

yudongqiu committed:
implement smirnoff_hack to cache ForceField creation

yudongqiu committed:
implement offxml_unit_strs in forcefield module to support v0.3

Commit:
Added slightly stricter convergence criteria.

yudongqiu committed:
improve sminoff error handling and engine optimize stability

Commit:
Relaxed the slightly stricter convergence criteria.

Commit:
Temporarily force simulations only.

yudongqiu committed:
increase maxIterations in simulation.minimizeEnergy

yudongqiu committed:
update opt_geo_target indicate to print name

leeping committed:
Merge pull request #146 from yudongqiu/improve_smirnoff

major improvements for smirnoff and optgeo targets

leeping committed:
Improved force field parameter representation and SMIRKS-based masking

leeping committed:
Use the OpenFF API for force field modification

leeping committed:
Merge pull request #147 from leeping/smirks_mask

Improved force field parameter representation and SMIRKS-based masking

leeping committed:
Remove unneeded calls to smirnoff_hack

leeping committed:
Merge pull request #148 from leeping/openff_api

Use the OpenFF API for force field modification

Commit:
Removed the simulation layer only constraint.

yudongqiu committed:
improve assign_openff_parameter performance by caching parameter search

Commit:
Added missing docstring.

Commit:
Merge pull request #2 from SimonBoothroyd/property_estimator_jacobian

Direct Gradient Calculations Through PropertyEstimator API

Commit:
Exposed options for how properties are estimated.
Properties are now loaded in as a data set.
Updated the targets folder.
Minor linting.

yudongqiu committed:
fix read_mvals issue with SMIRKs

Commit:
Removed verbose log statements.

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,340 @@
 
 try:
-    import property_estimator
+    # noinspection PyUnresolvedReferences
```
Are these specific comments for PyCharm editor you use? If so shall we remove them?

SimonBoothroyd replied:
They are (I think) - it probably makes sense to!

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -6,127 +6,340 @@
+        weights: list of PropertyWeight
+            The weighting of each property which will be optimised against.
+        denominators: list of PropertyDenominator
+            ???
```
We can use this description: "The denominators will be used to remove units from the properties and scale their values."

SimonBoothroyd replied:
Good suggestion!

yudongqiu commented on src/property_estimator_io.py:
```diff
@@ -142,59 +355,188 @@ def submit_jobs(self, mvals, AGrad=True, AHess=True):
+        if AGrad is False:
+            return estimated_data, estimated_gradients
+
+        jacobian = self._build_pvals_jacobian(mvals)
```
It's nice to see this implementation leads to using gradients in PE. I have three questions:
1. Are these gradients validated against the numerical gradients of mvals? If so, can we keep the numerical gradients code in here as a sanity check option?
2. The `pvals` does not contain all changing parameters in the ForceBalance parameterization. These are the "freely moving" ones. ForceBalance also supports "parameter_eval" and "parameter_repeat", and when those are defined the one-to-one mapping `_gradient_key_mappings` will lead to gradients of the evaluated parameters, (the ones not in pvals) to be lost. You can see [here](https://github.com/leeping/forcebalance/blob/91e3be696d91489d289ca13444435e9d6180b520/src/forcefield.py#L816), that `pfields` have all fields in the Force field file that is changing. And Lee-Ping recently implemented OpenFF [API calls](https://github.com/leeping/forcebalance/blob/91e3be696d91489d289ca13444435e9d6180b520/src/smirnoffio.py#L140) to modify OpenFF Forcefield object. (If you merge the master branch you will get them.) The point is we want to collect gradients for all `pfields` into `mvals`, not just `pvals`.
3. I think the `_build_pvals_jacobian()` gives the same result so shall we store that and skip re-computing it?

SimonBoothroyd replied:
1. Currently I'm comparing them against the values of the native FB implementation - my preference would be to use FB as the reference in this case (as opposed to the initial PE implementation) as it's a good sign when two independent code implementations are in good agreement!

2. Apologies I'm not sure that I understand this point fully - would you be able to give a few examples of cases where this would occur? 

   If these options aren't set frequently (or more pressing if they don't get set in the first round of fitting) it might be good for now to just raise a `NotImplementedException` in those cases, and revisit this at a later date just to help ease the time constraints in this first sprint (although if the effort to add this is minimal I'm open to trying to fix it now rather than postpone it until later 🙂)!

3. I think the latency of re-calculating the jacobian is well hidden by the cost of evaluating the physical properties so it shouldn't really contribute to the performance, but happy to add the caching if you'd prefer that?

yudongqiu replied:
For 2, I also had the same idea at first when implementing other components, but then I realized that we do need those features. An example is this test proposed by Mobley group: https://github.com/openforcefield/openforcefield/pull/384/files
If you look at line 300 in that file, the `parameter_eval` rule is set to use a special `sqrtk1` parameter to set the `k1` parameter. Here you will get the value of `sqrtk1` in the `pvals ` list, but actually `k1` is the "real" parameter. The `parameter_eval` feature can also be used to set relations between different terms, i.e. constraining `k1` of two different SMIRKs to be the same.

For 3, I agree the cost of evaluating the physical properties is higher than other parts of the code, so we designed the code to distribution the evaluation cost to many "worker nodes". In comparison, the `_build_pvals_jacobian` takes serial execution time on the host. Although it is cheap for small number of parameters, it's not exactly clear how mush time cost it will add eventually when the parameter relations are all evaluated in this function. Because this time cost can be easily eliminated by caching the result, I think it's better to do that.

yudongqiu replied:
For 1, after think more about it, I think I can understand your reason. The general idea is that, we have (a) ForceBalance liquid target as reference, and we implemented (b) numerical gradients, it matches the reference, then we implement (c) analytical gradients and it also matches the reference. I think you're saying as long as (c) matches (a), we don't need to keep (b) in the code.
I agree with this. However I think (b) is still valuable, because when (c) does not match (a), we will have a middle step to be able to tell which part goes wrong. We have found this very helpful, and you can see examples in the implementation of (a) where Lee-Ping still keep the FDCheck code just in case. https://github.com/leeping/forcebalance/blob/8cbafff745b416b513045b1eacbd58e14af22c16/src/data/npt.py#L526

j-wags replied:
> For 2, I also had the same idea at first when implementing other components, but then I realized that we do need those features. An example is this test proposed by Mobley group: https://github.com/openforcefield/openforcefield/pull/384/files

I was just summarizing the Torsions meeting, and I think we've decided to skip improper fitting for release-1. We are probably OK to have it raise a NotImplementedError if this comes along.

Commit:
Now prints the substance id.

yudongqiu commented:
I think this PR is very close to be done now! Great job.

Here are a few things left:

- [x] Include updated `forcebalance/studies/023_property_estimator/optimize.out`

- [ ] keep numerical mvals gradients option as a sanity check.

- [x] Try improve jacobian calculation to support `parameter_eval` and `parameter_repeat` features. If it's too much work, we can revisit this later, and the sanity check with numerical gradients will be very helpful validating this.

yudongqiu committed:
Merge pull request #149 from yudongqiu/improve_off_perf

improve assign_openff_parameter performance by caching parameter search

yudongqiu committed:
improve optgeo gradients

yudongqiu committed:
clean up opt_geo_target

yudongqiu committed:
remove 1/sqrt(n_bonds) terms

yudongqiu committed:
optgeo target convert all angle units radian to degree

yudongqiu committed:
update studies/022_opt_geo_target

yudongqiu committed:
improve work_queue in nifty.py

yudongqiu committed:
update default angle denominators to new unit

yudongqiu committed:
update prior widths in studies/022_opt_geo_target

yudongqiu committed:
Merge pull request #150 from yudongqiu/optgeo

Improve optgeo target

Commit:
Merge branch 'master' of https://github.com/leeping/forcebalance into property_estimator

Commit:
First pass at moving away from pvals.

Commit:
Removed Jacobian dependence on pvals. Ready for testing.

Commit:
Removed unused option.

Commit:
Updated the smirnoff force field to the new spec.
Temporarily disabled the ab initio target.

Commit:
Fixed missing default units for HVap.

Commit:
Minor fix for unitless gradients.

Commit:
Minor fix for unitless gradients.

Commit:
Added debug logging.

Commit:
Minor refactor to use pint.

Commit:
Updated the results of the study.

yudongqiu commented:
Although the numerical mvals gradients as a sanity check is a good feature to have, we agreed that we don't have time for re-adding that now, so we will revisit that later.

Everything else looks good and I will merge now.
